File: tests/python/test_config.py
"""Config tests aligned with the current direct-dict config API.

The historical ``get_*_config`` accessor helpers were removed in favour of plain
dict access (``config['learner']`` etc.). These tests validate ``load_config``
and ``_validate_config`` against the real ``config.toml`` and the current schema.
"""

import os
import sys
import tempfile

import pytest

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', 'python'))

from config import load_config, _validate_config  # noqa: E402

CONFIG_PATH = os.path.join(os.path.dirname(__file__), '..', '..', 'config.toml')

REQUIRED_SECTIONS = [
    'learner', 'training', 'model', 'features',
    'reward', 'data', 'logging', 'timeframes',
    'monitoring', 'per', 'backtesting', 'api',
]


@pytest.fixture
def config():
    return load_config(CONFIG_PATH)


class TestLoadConfig:
    def test_loads_valid_config(self, config):
        assert isinstance(config, dict)
        for section in REQUIRED_SECTIONS:
            assert section in config, f"missing section {section}"

    def test_raises_on_missing_file(self):
        with pytest.raises(FileNotFoundError):
            load_config('/nonexistent/path/config.toml')

    def test_device_resolved(self, config):
        # 'auto'/'cuda' are resolved to a concrete device by load_config.
        assert config['learner']['device'] in ('cuda', 'cpu')


class TestValidation:
    def test_raises_on_missing_section(self):
        tmp_path = None
        try:
            with tempfile.NamedTemporaryFile(mode='w', suffix='.toml', delete=False) as f:
                tmp_path = f.name
                f.write('[learner]\nhost = "tcp://127.0.0.1"\nport = 5555\n')
            with pytest.raises(ValueError, match="Missing required"):
                load_config(tmp_path)
        finally:
            if tmp_path and os.path.exists(tmp_path):
                os.unlink(tmp_path)

    def test_raises_on_no_actors(self):
        minimal = {s: {} for s in REQUIRED_SECTIONS}
        with pytest.raises(ValueError, match="at least one"):
            _validate_config(minimal)


class TestSchemaValues:
    def test_actors(self, config):
        symbols = [a['symbol'] for a in config['actors']]
        assert len(symbols) >= 1
        assert 'BTCUSDT' in symbols

    def test_training_invariants(self, config):
        t = config['training']
        assert 0.0 < t['gamma'] <= 1.0
        assert t['lr'] > 0
        assert t['batch_size'] > 0
        assert t['buffer_capacity'] >= t['min_buffer_size']
        assert t['epsilon_start'] >= t['epsilon_end']

    def test_model_features_consistency(self, config):
        assert config['model']['num_actions'] == 4
        assert config['features']['num_features'] > 0
        # d_model (= conv1d_filters) must be divisible by attention heads.
        assert config['model']['conv1d_filters'] % config['model']['n_attention_heads'] == 0

    def test_per_bounds(self, config):
        per = config['per']
        assert 0.0 <= per['alpha'] <= 1.0
        assert 0.0 <= per['beta_start'] <= per['beta_end'] <= 1.0

    def test_timeframes_non_negative(self, config):
        for k, v in config['timeframes'].items():
            assert v >= 0, f"{k} must be >= 0"
        assert any(v > 0 for v in config['timeframes'].values())

File: tests/node/test_config.js
const assert = require('assert');
const path = require('path');
const {
    loadConfig,
    getActorsConfig,
    getTimeframesConfig,
    getLearnerConfig,
    getTrainingConfig,
    getMonitoringConfig,
    getDataConfig,
    getRewardConfig,
    getFeaturesConfig,
} = require(path.join(__dirname, '..', '..', 'node', 'config'));

function test(name, fn) {
    try {
        fn();
        console.log(`  ✓ ${name}`);
    } catch (err) {
        console.log(`  ✗ ${name}`);
        console.log(`    ${err.message}`);
        process.exitCode = 1;
    }
}

console.log('Config Loader Tests (Node.js)\n');

const configPath = path.join(__dirname, '..', '..', 'config.toml');
let config;

test('loads valid config', () => {
    config = loadConfig(configPath);
    assert.ok(config && typeof config === 'object');
});

test('throws on missing file', () => {
    assert.throws(() => loadConfig('/nonexistent/config.toml'), /not found/);
});

// Getters return the matching sections (API contract, not brittle values that drift).
test('learner getter exposes connection fields', () => {
    const learner = getLearnerConfig(config);
    assert.strictEqual(learner, config.learner);
    for (const k of ['host', 'port', 'metrics_port', 'device']) {
        assert.ok(k in learner, `learner missing ${k}`);
    }
    assert.ok(typeof learner.device === 'string' && learner.device.length > 0);
});

test('actors getter returns at least one actor with a symbol', () => {
    const actors = getActorsConfig(config);
    assert.ok(Array.isArray(actors) && actors.length >= 1);
    assert.ok(actors.every(a => typeof a.symbol === 'string' && a.symbol.length > 0));
    assert.ok(actors.some(a => a.symbol === 'BTCUSDT'));
});

test('timeframes getter has all required keys and a positive base TF', () => {
    const tf = getTimeframesConfig(config);
    for (const k of ['candles_1m', 'candles_15m', 'candles_1h', 'candles_1d', 'candles_1w']) {
        assert.ok(k in tf, `timeframes missing ${k}`);
        assert.ok(tf[k] >= 0, `${k} must be >= 0`);
    }
    assert.ok(Object.values(tf).some(v => v > 0), 'at least one timeframe must be enabled');
});

test('training getter holds sane invariants', () => {
    const t = getTrainingConfig(config);
    assert.ok(t.gamma > 0 && t.gamma <= 1, `gamma=${t.gamma}`);
    assert.ok(t.lr > 0, `lr=${t.lr}`);
    assert.ok(t.batch_size > 0);
    assert.ok(t.buffer_capacity >= t.min_buffer_size, 'buffer_capacity must be >= min_buffer_size');
    assert.ok(t.epsilon_start >= t.epsilon_end, 'epsilon_start must be >= epsilon_end');
});

test('monitoring getter exposes ports', () => {
    const mon = getMonitoringConfig(config);
    assert.ok(Number.isInteger(mon.port));
    assert.ok(Number.isInteger(mon.metrics_pull_port));
});

test('data getter has source and positive normalization window', () => {
    const data = getDataConfig(config);
    assert.ok(typeof data.source === 'string');
    assert.ok(data.normalization_window > 0);
});

test('reward getter has commissions and a valid clip range', () => {
    const reward = getRewardConfig(config);
    assert.ok(reward.commission_open >= 0 && reward.commission_close >= 0);
    assert.ok(reward.clip_min < reward.clip_max, 'clip_min must be < clip_max');
});

test('features getter reports a positive feature count', () => {
    const features = getFeaturesConfig(config);
    assert.ok(features.num_features > 0);
});

console.log('\nDone.');
